### ISSUE TITLE
App crash on opening graphs near Montreal, CA

### ISSUE THREAD

**Issue by cbkiyanda** [martykan/forecastie#422]
The app crashes whenever opening the charts for cities around Montreal, Canada. Cities tested so far are
Montreal, CA
Longueuil, CA
Brossard, CA

The charts have worked fine brfore, so I think this is related
There is a forecast of heavy icing rain in the coming days in that region.

**Comment by cbkiyanda** [comment]
The day after, the behaviour is gone. Unfortunately, I didn't have time to download the corresponding file served by owm for further evaluation.

**Comment by robinpaulson** [comment]
When you say "heavy icing rain", I assume that is the same as sleet?

There has been discussion recently about the difference between rain and snow in the context of the graphs, I wonder if there's an edge case here we don't deal with correctly?

Thanks for bringing this up, could you keep an eye on this and post here if there are any further problems.

**Comment by FridoDeluxe** [comment]
Unfortunately `GraphActivity.java` has a sightly different JSON parser implementation than `MainActivity.java` (I wonder why there is more than one implementation, merging them could avoid such mistakes in the future).

The code for `GraphActivity.java` probably throws a NullPointerException in case that neither rain nor snow is included in the JSON response from OWM (see https://github.com/martykan/forecastie/blob/master/app/src/main/java/cz/martykan/forecastie/activities/GraphActivity.java#L345). Of course a real JSON example for such a weather condition would be helpful - we woud see how the JSON looks like and can veriy that it really causes the crash.

**Comment by robinpaulson** [comment]
Yes to merging parsers, that would reduce redundancy as you say.

Examples would definitely help. I'll cast around in a day or two, see if I can find a place with the conditions you suggest. @carminemassarelli it would help no end if you could answer my question above about "heavy icing rain".

**Comment by cbkiyanda** [comment]
With regards to the difference between icing rain and sleet. There's a bit of regionalisms between the different terms, but I would classify as follows (sorry for being pedantic and defining snow and rain as well. I'm a university professor...): 

**snow:** condensing water that nucleates to solids high in the atmosphere. Nuclei grow into crystal and falls as solids. Precipitation only encounters sub-freezing temperatures as it falls.
**Sleet**: condensing water turns to solid, encounters a region of warm temperatures which melts the precipitation back to liquid water. Close to the ground, the temperature is below freezing, which turns the water droplets back to solid **before** they hit the ground.
**icing rain** of **freezing rain**: As for sleet, the falling snow encounters a region of warm temperatures that melts it back to liquid. The difference between freezing rain and sleet is that in freezing rain, the warm layer extends closer to the ground and the water droplets don't have time to nucleate and turn into a solid again. The drops end up supercooled (temperature less than freezing, yet still liquid) and the nucleating to solid happens on contact with the ground or an object below freezing temperature. The ice forms on impact, covering surfaces with ice.
**rain** : as one expects, the water droplets stay in above freezing temperature and fall as liquid.

Reading around to figure out the translations of sleet to French, I found a relatively good plot on wikipedia representing the temperature difference of the different forms of precipitation:
https://en.wikipedia.org/wiki/Freezing_rain#/media/File:Ice_Storm_Chart.svg

In summary, snow and sleet hit the ground as solids, while freezing rain and rain hit the ground as liquids. Sleet is like "reformatted snow" in that it is snow that has been melted and re-frozen into tiny pebbles. It is distinct from hail, which is also solid pebbles of different sizes, but comes from a different mechanism. Sleet is formed because of a warm layer above a cold ground layer, while hail is formed because vertical wind patterns bring water droplets in higher atmosphere regions where it freezes. Hail does not typically happen in winter.
Here's a meteorologist explaining the difference between hail and sleet: https://www.youtube.com/watch?v=aO5qWZdW4FM

**Comment by carminemassarelli** [comment]
According to @cbkiyanda i found this difference which I report shortly: icing rain occurs when the snowflakes encountering warmer air become partially liquid maintaining the crystalline structure and then refreeze once in contact with the ground. Sleet, on the other hand, is a solid or mixed liquid-solid precipitation

**Comment by robinpaulson** [comment]
@cbkiyanda Thank you for the explanation, that is a lot to digest, I will think about it over the next day or so.

**Comment by FridoDeluxe** [comment]
> There has been discussion recently about the difference between rain and snow in the context of the graphs [...]

Just for the record: Currently the OWM response is checked for a `rain` value, if this is not available then `snow` is taken. The value is then saved as _rain_ (variable names as well as UI strings). So when Forecastie talks about rain mostly precipitation in general is meant.

**Comment by FridoDeluxe** [comment]
The above pull request should mitigate potential NullPointerException when both rain and snow is missing in the OWM response. Hopefully it fixes this issue.

Still an actual JSON response would be helpful. And we should consolidate the parsing code into one place eventually. (Or evaluate if it's better to replace our parsing code by a dedicated library like this one: https://github.com/iamashks/OWM-JAPIs)

**Comment by robinpaulson** [comment]
Cheers for the PR. That library does look interesting and very tempting. We really ought to abstract it out somehow.

**Comment by robinpaulson** [comment]
@cbkiyanda As an interim, I've released a quick fix to prevent crashing in the circumstances you had. It's not a full fix, but it will prevent the app crashing. We'll release a fuller fix when we figure out what's going on; more investigation of the OWM API is required. The new version will be out in 1-4 days, as per F-Droid updates.

I'll keep this issue open until it's completely fixed.

**Comment by cbkiyanda** [comment]
Thanks very much. I will keep an eye out for the update and will keep testing to see if the bug or a similar one is triggered.

I have been meaning to learn the basics of android app development for a while now, though time has been lacking. If I can find some, I might tackle the presentation of the precipitation graph. Unfortunately, my coding abilities are mostly about numerical computing in fortran, c, python, so there's a bit of a learning curve on a project like this.

**Comment by cbkiyanda** [comment]
I've just gotten the same behaviour as initially reported. I'm running 1.13. I haven't seen the recent update come through F-droid. This time, I did manage to get the json forecast file from owm.

[forecast.zip](https://github.com/martykan/forecastie/files/4108897/forecast.zip)

**Comment by FridoDeluxe** [comment]
@cbkiyanda Great, thanks for the JSON file. Seems like different cases are possible when receiving data from OWM:
- rain/snow object is completely missing
- rain/snow object is there, but empty: `{}`
- rain/snow contains a value for 3h or 1h

I also took a closer look again at the code and it looks like my PR #427 was not necessary, because missing or empty percipitation values should be handled in `MainActivity.getRainString()`. Running the GraphActivity with the old code (version 1.13) and your JSON example couldn't reproduce the error. Either the JSON is not exactly the same JSON that caused the crash, or the real reason for the bug is something else ... It would be nice if you could also provide the logcat output the next time it crashes?

**Comment by robinpaulson** [comment]
v1.13.1 and .2 are taking an unusually long time to be published. It's likely .1 will be skipped altogether and go straight to .2

Regarding the error, I've noticed previously we don't trap partial download errors so well, perhaps that is to blame? We're not great at trapping errors/showing error messages full stop. See #316 for more.
